### PR TITLE
Improved crash dialog

### PR DESCRIPTION
**Pull request by NathanW2**
## Description
Includes improvements to include a stack trace and report details in the crash dialog.  The stack trace is of course only goig to work on debug based builds (RelWithDebInfo or Debug) and PDB files but we do both those on OSGeo4W at least so that is ok.  There is also a symbol server Jurgen setup we can use post release if we need,

Part of this is also generating a crash ID from the stack frames.  This can be used to group crashes when we collect them in for reporting, could also be used to let the user know in the crash dialog if it's already been fixed.

At the moment the crash ID is a hash of all QGIS based stack lines, however, this increases the entropy so we might want to only pull the first 2 to mark a single crash point instead of the whole stack.

image

**Note: Not for merge as I have reverted another commit to just get a build working :)**

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by NathanW2** [comment]
The revet kills the diff as expected so let me rebase and clean that up.

File: src/app/qgscrashdialog.cpp [hunk 1]
```diff
@@ -17,16 +17,42 @@
...
   : QDialog( parent )
 {
   setupUi( this );
-  setWindowTitle( tr( "Oh. Snap!" ) );
+  setWindowTitle( tr( "Oh Uh!" ) );
 
-  mCrashHeaderMessage->setText( tr( "Ouch!" ) );
-  mCrashMessage->setText( tr("Sorry it looks like QGIS crashed. \nSomething unexpected happened that we didn't handle correctly."));
+  mCrashHeaderMessage->setText( tr( "Oh Uh!" ) );
+  mCrashMessage->setText( tr( "Sorry. It looks something unexpected happend that we didn't handle and QGIS crashed." ) );
```
Comment (nyalldawson) [inline src/app/qgscrashdialog.cpp]: happend  -> happened

File: src/app/qgscrashdialog.cpp [hunk 2]
```diff
@@ -17,16 +17,42 @@
...
+  connect( mCopyReportButton, &QPushButton::clicked, this, &QgsCrashDialog::createBugReport );
 
-  mHelpLabel->setText( tr( "Keen to help us fix it? <br> <a href=\"http://qgis.org/en/site/getinvolved/development/bugreporting.html#bugs-features-and-issues\"> Follow the steps to help devs.<a>" ) );
+  mHelpLabel->setText( tr( "Keen to help us fix bugs?"
+                           "<br> <a href=\"http://qgis.org/en/site/getinvolved/development/bugreporting.html#bugs-features-and-issues\">Follow the steps to help our developers.</a>"
+                           "<br><br>"
+                           "You can also send us a helpful bug report using the Copy Report button"));
```
Comment (nyalldawson) [inline src/app/qgscrashdialog.cpp]: button -> button.

File: src/app/qgscrashreport.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    {
+      reportData.append( QString( "QGIS code branch: Release %1.%2" )
+                         .arg( Qgis::QGIS_VERSION_INT / 10000 ).arg( Qgis::QGIS_VERSION_INT / 100 % 100 ) );
+    }
+    else
+    {
+      reportData.append( QString( "QGIS code revision: %1" ).arg( Qgis::QGIS_DEV_VERSION ) );
+    }
+
+    reportData.append( QString( "Compiled agaist Qt: %1" ).arg( QT_VERSION_STR ) );
```
Comment (nyalldawson) [inline src/app/qgscrashreport.cpp]: agaist  -> against

File: src/app/qgscrashreport.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    reportData.append( QString( "Running agaist GDAL: %1" ).arg( GDALVersionInfo( "RELEASE_NAME" ) ) );
+  }
+
+  if ( flags().testFlag( QgsCrashReport::SystemInfo ) )
+  {
+    reportData.append( "\n" );
+    reportData.append( "System Info" );
+    reportData.append( "===========" );
+    reportData.append( QString( "CPU Type: %1" ).arg( QSysInfo::currentCpuArchitecture() ) );
+    reportData.append( QString( "Kernal Type: %1" ).arg( QSysInfo::kernelType() ) );
```
Comment (nyalldawson) [inline src/app/qgscrashreport.cpp]: Kernal  -> Kernel

File: src/app/qgscrashreport.h [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      QString ModuleName;
+      QString SymbolName;
+      QString FileName;
+      QString LineNumber;
+
+      /**
+       * Check if this stack line is part of QGIS.
+       * @return True if part of QGIS.
+       */
+      bool IsQGISModule();
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: -> isQgisModule() const

File: src/app/qgscrashreport.h [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+      /**
+       * Check if this stack line is valid.  Considered valid when the filename and line
+       * number are known.
+       * @return True of the line is valid.
+       */
+      bool IsValid();
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: -> isValid()

File: src/app/qgscrashreport.h [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      /**
+       * Check if this stack line is part of QGIS.
+       * @return True if part of QGIS.
+       */
+      bool IsQGISModule();
+
+      /**
+       * Check if this stack line is valid.  Considered valid when the filename and line
+       * number are known.
+       * @return True of the line is valid.
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: Should be \return now - no more @ doxygen tags

File: src/app/qgscrashreport.h [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+     * Sets the stack trace for the crash report.
+     * @param value A string list for each line in the stack trace.
+     */
+    void setStackTrace( const QList<QgsCrashReport::StackLine> value ) { mStackTrace = value; }
+
+    /**
+     * Returns the stack trace for this report.
+     * @return A string list for each line in the stack trace.
+     */
+    const QList<QgsCrashReport::StackLine> StackTrace() { return mStackTrace; }
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: Can you change this to:

    QList<QgsCrashReport::StackLine> StackTrace() const 

(ie const at end not start)

File: src/app/qgscrashreport.h [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+      SystemInfo           = 6,
+      QGISInfo             = 8,
+    };
+    Q_DECLARE_FLAGS( Flags, Flag )
+
+    /**
+     * Sets the stack trace for the crash report.
+     * @param value A string list for each line in the stack trace.
+     */
+    void setStackTrace( const QList<QgsCrashReport::StackLine> value ) { mStackTrace = value; }
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: missing '&'

File: src/app/qgscrashreport.h [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+     * Returns the include flags that have been set for this report.
+     * @return The flags marking what details are included in this report.
+     */
+    const Flags &flags() const { return mFlags; }
+
+    /**
+     * Generate a string version of the report.
+     * @return A formatted string including all the information from the report.
+     */
+    QString toString();
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: const

File: src/app/qgscrashreport.h [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+    /**
+     * Generates a crash ID for the crash report.
+     * @return
+     */
+    QString crashID();
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: const

**Review by nyalldawson — COMMENTED** [review]
Just some general comments - can you unrevert the keychain commit as it's making this hard to review?

**Comment by NathanW2** [comment]
@nyalldawson removed the commit.

File: src/app/qgscrashreport.cpp [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+  QString report;
+  Q_FOREACH ( QString line, reportData )
+  {
+    report += line + "\n";
+  }
+  return report;
+}
+
+QString QgsCrashReport::crashID()
```
Comment (NathanW2) [inline src/app/qgscrashreport.cpp]: @nyalldawson @m-kuhn @wonder-sk keen for your thoughts on this method. I want generate a crash ID but maybe going all the way up the stack adds to much chance for things to change.  Maye just 2 or 3 frames is enough.

File: src/app/qgscrashreport.h (hunk 8 above)
Comment (NathanW2) [inline src/app/qgscrashreport.h]: This is why coding in the morning on the train is a bad idea.

File: src/app/qgscrashreport.cpp (hunk 12 above)
Comment (wonder-sk) [inline src/app/qgscrashreport.cpp]: I think I would prefer ID out of the full stack trace... often top of the trace may be something inside Qt code and we may wrongly identify different crashes as the same one. I would not be worried that things would be too easy to change and thus produce different crash IDs for the same problem - I guess most of the time users will use one of the few released versions where the crash IDs should be stable...

**Comment by NathanW2** [comment]
At the moment I'm only including qgis modules does that sound alright?

File: src/app/qgisapp.cpp [hunk 13]
```diff
@@ -12506,45 +12507,98 @@ void QgisApp::transactionGroupCommitError( const QString &error )
...
+      QString lineNumber;
+      QString moduleName;
+      if ( SymGetLineFromAddr( process, ( DWORD )( stack_frame.AddrPC.Offset ), &dwDisplacement, line ) )
+      {
+        fileName = QString( line->FileName );
+        lineNumber = QString::number( line->LineNumber );
+      }
+      else
+      {
+        fileName = "(unkown file)";
```
Comment (jef-n) [inline src/app/qgisapp.cpp]: unknown

File: src/app/qgscrashreport.cpp (hunk 12 above)
Comment (NathanW2) [inline src/app/qgscrashreport.cpp]: Ok that's fine.  At the moment I'm only including QGIS modules but maybe we just do everything.

File: src/app/qgscrashdialog.h [hunk 14]
```diff
@@ -33,6 +38,12 @@ class APP_EXPORT QgsCrashDialog : public QDialog, private Ui::QgsCrashDialog
      * A dialog to show a nicer crash dialog to the user.
      */
     QgsCrashDialog( QWidget *parent = nullptr );
+
+    void setBugReport( QString reportData );
```
Comment (nyalldawson) [inline src/app/qgscrashdialog.h]: const QString &

File: src/app/qgscrashreport.cpp [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+{
+  QString data = QString::null;
+
+  Q_FOREACH ( QgsCrashReport::StackLine line, mStackTrace )
```
Comment (nyalldawson) [inline src/app/qgscrashreport.cpp]: Q_FOREACH ( const QgsCrashReport::StackLine &line, mStackTrace )

File: src/app/qgscrashreport.h [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    QgsCrashReport();
+
+  public:
+    enum Flag
+    {
+      Stack                = 1,
+      Plugins              = 2,
+      ProjectDetails       = 4,
+      SystemInfo           = 6,
+      QGISInfo             = 8,
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: QGISInfo -> QgisInfo

File: src/app/qgscrashreport.h [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+    /**
+     * Include information to generate user friendly crash report for QGIS.
+     */
+    QgsCrashReport();
+
+  public:
+    enum Flag
+    {
+      Stack                = 1,
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: These numbers aren't right - they need to be unique bits. Use

    Stack = 1 << 0,
    Plugins = 1 << 1,
    ProjectDetails = 1 << 2,
    SystemInfo = 1 << 3 (8, not 6)
etc

File: src/app/qgscrashreport.h [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+     * Set the flags to mark which features are included in this crash report.
+     * @param flags The flag for each feature.
+     */
+    void setFlags( QgsCrashReport::Flags flags );
+
+    /**
+     * Returns the include flags that have been set for this report.
+     * @return The flags marking what details are included in this report.
+     */
+    const Flags &flags() const { return mFlags; }
```
Comment (nyalldawson) [inline src/app/qgscrashreport.h]: Flags is just an int - no need to return a reference here.

**Comment by NathanW2** [comment]
Merged now.  Will refactor some of this out more to make it easier to port on other platforms.

**Comment by luipir** [comment]
hi @NathanW2 I¡m not used to debug in Win env. I had a crash report that you can see in: https://issues.qgis.org/issues/16890
this crash report is done installing qgis-dev and qgis-dev-pdb

btw I'no symbols resolved... how to setup OSGeo4W to allow symbols resolved in crash report? Probably can answer to this question also @jef-n

**Comment by luipir** [comment]
checking on linux what geometry is passed to:
https://github.com/qgis/QGIS/blob/master/src/core/qgsvectorlayereditbuffer.cpp#L329

**Pull request merged**